ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/Cache/IProvidedEntitiesCache.cs

```diff
@@ -1,67 +1,88 @@
-    T[] GetOrCreateBatch(TKey[] keys, int typeProviderId, ProvidedTypeContextHolder context);
+    T[] GetOrCreateBatch(TKey[] keys, int typeProviderId, TParam parameters);
 
     void Remove(int typeProviderId);
 
     string Dump();
   }
 
-  public abstract class ProvidedEntitiesCacheBase<T> : IProvidedEntitiesCache<T, int> where T : class
+  public abstract class ProvidedEntitiesCacheBase<T, TKey, TParam> : IProvidedEntitiesCache<T, TKey, TParam>
+    where T : class
   {
     protected readonly TypeProvidersContext TypeProvidersContext;
-    protected readonly IDictionary<int, T> Entities;
-    protected readonly IDictionary<int, List<int>> EntitiesPerProvider;
+    protected readonly ConcurrentDictionary<TKey, T> Entities;
+    protected readonly IDictionary<int, List<TKey>> EntitiesPerProvider;
+    private readonly object myEntitiesPerProviderLockObj = new object();
+    private SpinWaitLock myLock;
 
     protected ProvidedEntitiesCacheBase(TypeProvidersContext typeProvidersContext)
     {
       TypeProvidersContext = typeProvidersContext;
-      Entities = new Dictionary<int, T>();
-      EntitiesPerProvider = new Dictionary<int, List<int>>();
+      Entities = new ConcurrentDictionary<TKey, T>();
+      EntitiesPerProvider = new Dictionary<int, List<TKey>>();
     }
 
-    public T GetOrCreate(int key, int typeProviderId, ProvidedTypeContextHolder context)
+    public T GetOrCreate(TKey key, int typeProviderId, TParam parameters)
     {
-      if (key == ProvidedConst.DefaultId) return null;
+      if (!KeyHasValue(key)) return null;
       if (Entities.TryGetValue(key, out var providedEntity)) return providedEntity;
 
-      providedEntity = Create(key, typeProviderId, context);
-      AttachToTypeProvider(typeProviderId, key, providedEntity);
-
-      return providedEntity;
+      try
+      {
+        myLock.Enter();
+        if (Entities.TryGetValue(key, out providedEntity)) return providedEntity;
+        providedEntity = Create(key, typeProviderId, parameters);
+        AttachToTypeProvider(typeProviderId, key, providedEntity);
+        return providedEntity;
+      }
+      finally
+      {
+        myLock.Exit();
+      }
     }
 
-    public T[] GetOrCreateBatch(int[] keys, int typeProviderId, ProvidedTypeContextHolder context)
+    public T[] GetOrCreateBatch(TKey[] keys, int typeProviderId, TParam parameters)
     {
+      if (keys.Length == 1) return new[] { GetOrCreate(keys[0], typeProviderId, parameters) };
+
       var entities = new T[keys.Length];
 
       var groups = keys
         .Select((key, i) => (key, i))
-        .GroupBy(t => t.key == ProvidedConst.DefaultId || Entities.ContainsKey(t.key));
+        .GroupBy(t => !KeyHasValue(t.key) || Entities.ContainsKey(t.key));
 
       foreach (var group in groups)
       {
         if (group.Key)
           foreach (var (key, i) in group)
-            entities[i] = GetOrCreate(key, typeProviderId, context);
+          {
+            if (!KeyHasValue(key)) return null;
```

Comment:
This is very strange. Should this really return `null` as `T[]` from `GetOrCreateBatch`? Thought it should just do `continue`?

Reply:
you are right, there is a logical typo here